File: src/main/java/ru/reosfire/wmlib/yaml/common/wrappers/text/ClickConfig.java
package ru.reosfire.wmlib.yaml.common.wrappers.text;

import net.md_5.bungee.api.chat.ClickEvent;
import org.bukkit.OfflinePlayer;
import org.bukkit.configuration.ConfigurationSection;
import ru.reosfire.wmlib.text.IColorizer;
import ru.reosfire.wmlib.text.Replacement;
import ru.reosfire.wmlib.text.Text;
import ru.reosfire.wmlib.yaml.YamlConfig;
import ru.reosfire.wmlib.yaml.common.wrappers.WrapperConfig;

public class ClickConfig extends YamlConfig implements WrapperConfig<ClickEvent>
{
    public final ClickEvent.Action Action;
    public final String Value;
    public ClickConfig(ConfigurationSection configurationSection)
    {
        super(configurationSection);
        Action = ClickEvent.Action.valueOf(getString("Action"));
        Value = getColoredString("Value");
    }

    public ClickEvent unwrap(OfflinePlayer player, Replacement... replacements)
    {
        return unwrap(s -> Text.colorize(player, s, replacements));
    }
    public ClickEvent unwrap(Replacement... replacements)
    {
        return unwrap(s -> Replacement.set(s, replacements));
    }

    @Override
    public ClickEvent unwrap()
    {
        return new ClickEvent(Action, Value);
    }

    public ClickEvent unwrap(IColorizer colorizer)
    {
        return new ClickEvent(Action, colorizer.colorize(Value));
    }
}

File: src/main/java/ru/reosfire/wmlib/yaml/common/wrappers/text/ClickConfig.kt
package ru.reosfire.wmlib.yaml.common.wrappers.text

import net.md_5.bungee.api.chat.ClickEvent
import org.bukkit.OfflinePlayer
import org.bukkit.configuration.ConfigurationSection
import ru.reosfire.wmlib.extensions.colorize
import ru.reosfire.wmlib.extensions.setReplacements
import ru.reosfire.wmlib.text.IColorizer
import ru.reosfire.wmlib.text.Replacement
import ru.reosfire.wmlib.yaml.YamlConfig
import ru.reosfire.wmlib.yaml.common.wrappers.WrapperConfig

class ClickConfig(configurationSection: ConfigurationSection?) : YamlConfig(configurationSection),
    WrapperConfig<ClickEvent?> {
    val Action = ClickEvent.Action.valueOf(getString("Action")!!)
    val Value = getColoredString("Value")

    fun unwrap(player: OfflinePlayer, vararg replacements: Replacement): ClickEvent {
        return unwrap{ it?.colorize(player, *replacements) }
    }

    fun unwrap(vararg replacements: Replacement): ClickEvent {
        return unwrap{ it?.setReplacements(*replacements) }
    }

    override fun unwrap(): ClickEvent {
        return ClickEvent(Action, Value)
    }

    fun unwrap(colorizer: IColorizer): ClickEvent {
        return ClickEvent(Action, colorizer.colorize(Value))
    }
}